pkg/controller/appdefinition/checkimageallowed.go

```diff
@@ -0,0 +1,65 @@
+package appdefinition
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	v1 "github.com/acorn-io/acorn/pkg/apis/internal.acorn.io/v1"
+	"github.com/acorn-io/acorn/pkg/condition"
+	"github.com/acorn-io/acorn/pkg/imageallowrules"
+	"github.com/acorn-io/baaah/pkg/router"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/sirupsen/logrus"
+)
+
+// ImageAllowed is a middleware that checks if the image is allowed by the image allow rules.
+func ImageAllowed(h router.Handler) router.Handler {
+	return router.HandlerFunc(func(req router.Request, resp router.Response) error {
+		appInstance := req.Object.(*v1.AppInstance)
+		for _, cond := range appInstance.Status.Conditions {
+			if cond.Type == v1.AppInstanceConditionImageAllowed && !cond.Success {
+				return fmt.Errorf("image %s is not allowed by rule %s", appInstance.Spec.Image, cond.Reason)
+			}
+		}
+		return h.Handle(req, resp)
+	})
+}
+
+// CheckImageAllowedHandler is a router handler that checks if the image is allowed by the image allow rules and sets a status field accordingly
+func CheckImageAllowedHandler(transport http.RoundTripper) router.HandlerFunc {
+	return func(req router.Request, resp router.Response) error {
+
+		appInstance := req.Object.(*v1.AppInstance)
+		cond := condition.Setter(appInstance, resp, v1.AppInstanceConditionImageAllowed)
+
+		targetImage, unknownReason := determineTargetImage(appInstance)
+		if targetImage == "" {
+			if appInstance.Status.AppImage.Name != "" {
+				targetImage = appInstance.Status.AppImage.Name
+			} else {
+				if unknownReason != "" {
+					cond.Unknown(unknownReason)
+				} else {
+					cond.Error(fmt.Errorf("no image specified"))
+				}
+				return nil
```

Comment:
What situation would lead to this return? Do we want to be a message on the condition in this case?

Reply:
Since `appInstance.Status.AppImage.Name == ""` means that the image was not pulled/applied to the app yet, we would only end up here, if in addition to that the determination logic cannot find a suitable image, which is e.g. when the autoupgrade sync was never triggered before (see https://github.com/acorn-io/acorn/blob/main/pkg/controller/appdefinition/pullappimage.go#L77)

It's pretty unlikely though, except for something went really wrong.
I think though that we could never hit the `else` clause here.

Reply:
OK. That makes sense. Thank you for the explanation.

Do you think a message should be added to the condition just in case? Or is that a waste?